bpf/xdp_udp.bpf.c

```diff
@@ -0,0 +1,220 @@
+// +build ignore
+
+#include <linux/bpf_common.h>
+#include <linux/if_ether.h>
+#include <linux/in.h>
+#include <linux/ip.h>
+#include <linux/udp.h>
+#include <linux/bpf.h>
+#include <bpf_helpers.h>
+#include <bpf_endian.h>
+
+char __license[] SEC("license") = "GPL";
+
+#ifndef memcpy
+ #define memcpy(dest, src, n) __builtin_memcpy((dest), (src), (n))
+#endif
+
+#define MAX_BACKENDS 128
+#define MAX_UDP_LENGTH 1480
+
+#define UDP_PAYLOAD_SIZE(x) (unsigned int)(((bpf_htons(x) - sizeof(struct udphdr)) * 8 ) / 4)
+
+static __always_inline void ip_from_int(__u32 *buf, __be32 ip) {
+    buf[0] = (ip >> 0 ) & 0xFF;
+    buf[1] = (ip >> 8 ) & 0xFF;
+    buf[2] = (ip >> 16 ) & 0xFF;
+    buf[3] = (ip >> 24 ) & 0xFF;
+}
+
+static __always_inline void bpf_printk_ip(__be32 ip) {
+    __u32 ip_parts[4];
+    ip_from_int((__u32 *)&ip_parts, ip);
+    bpf_printk("%d.%d.%d.%d", ip_parts[0], ip_parts[1], ip_parts[2], ip_parts[3]);
+}
+
+static __always_inline __u16 csum_fold_helper(__u64 csum) {
+    int i;
+#pragma unroll
+    for (i = 0; i < 4; i++)
+    {
+        if (csum >> 16)
+            csum = (csum & 0xffff) + (csum >> 16);
+    }
+    return ~csum;
+}
+
+static __always_inline __u16 iph_csum(struct iphdr *iph) {
+    iph->check = 0;
+    unsigned long long csum = bpf_csum_diff(0, 0, (unsigned int *)iph, sizeof(struct iphdr), 0);
+    return csum_fold_helper(csum);
+}
+
+static __always_inline __u16 udp_checksum(struct iphdr *ip, struct udphdr * udp, void * data_end) {
+    udp->check = 0;
+
+    // So we can overflow a bit make this __u32
+    __u32 csum_total = 0;
+    __u16 csum;
+    __u16 *buf = (void *)udp;
+
+    csum_total += (__u16)ip->saddr;
+    csum_total += (__u16)(ip->saddr >> 16);
+    csum_total += (__u16)ip->daddr;
+    csum_total += (__u16)(ip->daddr >> 16);
+    csum_total += (__u16)(ip->protocol << 8);
+    csum_total += udp->len;
+
+    // The number of nibbles in the UDP header + Payload
+    unsigned int udp_packet_nibbles = UDP_PAYLOAD_SIZE(udp->len);
+
+    // Here we only want to iterate through payload 
+    // NOT trailing bits
+    for (int i = 0; i <= MAX_UDP_LENGTH; i += 2) {
+        if (i > udp_packet_nibbles) {
```

Comment:
Sorry I'm still getting confused here... maybe it's because i'm using the term nibble incorrectly

So the issue is, `data_end` is a PTR which includes the payload AND any trailing bits, while I forced `udp_packet_nibbles` to be the number of "digits" in the raw hex payload... 

For example in the following packet 

```bash
13:23:15.756911 06:56:87:ec:fd:1f > 86:ad:33:29:ff:5e, ethertype IPv4 (0x0800), length 60: (tos 0x0, ttl 57, id 20891, offset 0, flags [DF], proto UDP (17), length 33)
    10.8.125.12.58980 > 192.168.10.2.sapv1: [bad udp cksum 0xd301 -> 0xaf43!] UDP, length 5
        0x0000:  86ad 3329 ff5e 0656 87ec fd1f 0800 4500
        0x0010:  0021 519b 4000 3911 9e72 0a08 7d0c c0a8
        0x0020:  0a02 e664 2693 000d d301 7465 7374 0a00
        0x0030:  0000 0000 d2f2 935d 0000 0000
```

The UDP header + Data is 

```bash
e664 2693 000d d301 7465 7374 0a00
0000 0000 d2f2 935d 0000 0000
```

so 
```
    for (int i = 0; i <= MAX_UDP_LENGTH; i += 2) {
        if ((void *)(buf + 1) > data_end) {
            break;
        }
        csum_total += *buf;
        buf++;
    }
```

would iterate through all of that ^^ (i.e increment buff and read 13 times == 26 bytes of data)

Here packet length is 000d(13 bytes) so I know that I want to exit after incrementing buff 6.5 times ignoring those last 6.5 iterations (13 bytes of data)

The confusing part here is that `i` is a nibble index (1 hex digit (4 bytes)) (maybe) here, so that we can represent those 6 16 bit iterations and 1 8 bit iteration

I guess I'm confused trying to calculate 6 from the packet data

Reply:
A nibble is half a byte, i.e. 4 bits, 0..15 (what I think you are calling a hex digit). It's not useful to think about nibbles or try and do anything in terms of nibbles, except if you are writing your own code to convert numbers to hex strings.

```
__u16 *buf = (void *)udp;
```

This declares buf to be a pointer that will point to `__u16` width values. `buf++` will increment it by 2, to point at the next `__u16` width value:

```
__u16 *buf = 0xf00;
printf("%x\n", buf);
printf("%x\n", buf + 1);
f00
f02
```

This means that `if ((void *)(buf + 1) > data_end)` would trigger too early because `buf + 1` is actually `+2` and would exit even if there were 2 bytes to be read. 

It's safer to do all the arithmetic in bytes since the payload is a byte stream and the udp->len is in bytes.

Reply:
I updated this to be much simplier